include/kframework/lemmas/bytes-simplification.k

```diff
@@ -7,88 +7,240 @@ module BYTES-SIMPLIFICATION [symbolic]
   // Buffer Reasoning
   // ########################
 
-    rule 0 <=Int lengthBytes ( _ ) => true [smt-lemma, simplification]
+    rule [bytes-not-equal-length]:
+      BA1:Bytes ==K BA2:Bytes => false
+      requires lengthBytes(BA1) =/=Int lengthBytes(BA2)
+      [simplification]
 
-    rule BA:Bytes   ==K #buf(32, DATA) => #buf(32, DATA) ==K              BA                                    [simplification, concrete(BA)]
-    rule #buf(32, DATA) ==K BA:Bytes   =>          DATA  ==Int #asInteger(BA) requires lengthBytes(BA) <=Int 32 [simplification, concrete(BA)]
+    rule [bytes-equal-concat-split-ml]:
+      { A:Bytes +Bytes B:Bytes #Equals C:Bytes +Bytes D:Bytes } => { A #Equals C } #And { B #Equals D }
+      requires lengthBytes(A) ==Int lengthBytes(C)
+      [simplification]
 
-    rule #range(M, N1 +Int N2, K) => #range(#range(M, N2, lengthBytes(M) -Int N2), N1, K)
-      requires 0 <=Int N1 andBool 0 <=Int N2
-      [simplification, concrete(N2), concrete(M)]
+    rule [bytes-concat-empty-right]: B:Bytes +Bytes .Bytes  => B [simplification]
+    rule [bytes-concat-empty-left]:   .Bytes +Bytes B:Bytes => B [simplification]
 
-    rule #range(A +Bytes _B, START, LENGTH) => #range(A, START, LENGTH)
-      requires START +Int LENGTH <=Int lengthBytes(A) [simplification]
+    rule [bytes-concat-right-assoc]: (B1:Bytes +Bytes B2:Bytes) +Bytes B3:Bytes => B1 +Bytes (B2 +Bytes B3) [simplification]
 
-    rule #range(A +Bytes B, START, LENGTH) => #range(B, START -Int lengthBytes(A), LENGTH)
-      requires lengthBytes(A) <=Int START [simplification]
+    // #buf
 
-    rule #range(BA:Bytes, 0, LEN) => BA requires lengthBytes(BA) ==Int LEN [simplification]
+    rule [buf-ceil]:
+      #Ceil(#buf ( X , _:Int )) => 0 <=Int X
+      [simplification]
```

Comment:
This rule is not allowed by new rules coming down, and likely is not actually meaning what you think it is: https://github.com/runtimeverification/k/issues/3284 and https://github.com/runtimeverification/k/issues/3286 for reference.

You may be able to get away with `#Ceil(#buf(W, _:Int)) => #Bottom requires W <Int 0 [simpilification]`, and `#Ceil(#buf(W, I)) => #Bottom requires 2 ^Int (8 *Int W) <=Int I`, for example, while still adhering to the rules linked above.

But recall that `#Ceil(...)` returns `#Top` or `#Bottom`, which is either the set `{true, false} == #Top`, or `{} = #Bottom`. So saying that `#Top => {true}` or `#Top => {false}` or `#Bottom => {true}` or `#Bottom => {false}` is a bit non-sensical in matching-logic, and is equivalent to adding an axiom which says "sort `Bool` only has one element".

Reply:
Yes yes, that rule should not be allowed, what I meant to write was `{ true #Equals 0 <=Int X }`, is that the correct way?

Reply:
Actually, I'm not sure what's happening, `#buf` is declared as `total` in `buf.md`, so there should be no `#Ceil` issues, but there are some constraints in place on how it should be used.

Reply:
I've removed this simplification in the latest commit, as `#buf` is declared as total